FSDP optimizer overlap

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104813
* __->__ #98667


constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

rohan-varma committed:
[WIP] Prototype FSDP optimizer overlap

constraints:

1. No support for gradient accumulation
2. CPU offload not supported
3. Step is waited on in post backward final cb, when in theory it can wait
until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Prototype FSDP optimizer overlap"


constraints:

1. No support for gradient accumulation
2. CPU offload not supported
3. Step is waited on in post backward final cb, when in theory it can wait
until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Prototype FSDP optimizer overlap"


constraints:

1. No support for gradient accumulation
2. CPU offload not supported
3. Step is waited on in post backward final cb, when in theory it can wait
until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP optimizer overlap"


constraints:

1. No support for gradient accumulation
2. CPU offload is not performant as the step still occurs on CPU. Will fix this for large scale CPU offload use cases as follow up.
3. Step is waited on in post backward final cb, when in theory it can wait
until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -811,6 +824,20 @@ def _post_backward_hook(
+                        ):
+                            for optim in orig_param._in_backward_optimizers:
+                                optim.step()
+                            # Not sure if we need to do this. Setting entire
```
In my understanding, we should still do this.

We do not any references keeping the flat gradient alive, and we want the invariant that "if `flat_param.grad is None`, then each original parameter's `.grad is None`".

awgu commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -442,6 +442,9 @@ def __init__(
         self._training_state = HandleTrainingState.IDLE
         self._debug_level = dist.get_debug_level()
         self._fully_sharded_module = fully_sharded_module
+        # NB: this is updated in the first iteration because optimizers can be applied after
```
nit: (in case lazy init happens due to something like loading state dict and not running first forward)
```suggestion
        # NB: this is updated during lazy initialization because optimizers can be applied after
```

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -912,6 +939,9 @@ def _post_backward_final_callback(
 
     if root_state._sync_gradients:
+        # TODO: also waits for optimizer step to finish. This can be pushed to the next forward
+        # by recording an event for each individually wrapped FSDP module, and only waiting
```
In order for this mentioned solution to work, do we need to run the overlapped optimizer outside the post-backward stream and in a separate stream? Otherwise, `wait_stream()` is too coarse grained to differentiate between the optimizer computation and other existing post-backward ops.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -416,6 +425,10 @@ def _pre_forward(
         # their gradients. They must be re-registered every forward pass in case
         # the `grad_fn` is mutated.
         _register_post_backward_hooks(state, handles)
+        # We may have to reallocate the _cpu_grad if optimizer overlap set the grad to None in the backward pass.
```
Is this true? If we set `.grad` to `None`, the `._cpu_grad` still exists?

rohan-varma replied:
@awgu I'm keeping this since I'm setting cpu_grad=None in the post-backward. This is to maintain the invariant that if all orig param grad = None, then handle.sharded_grad should be None. In this case (when CPU offload enabled), handle.sharded_grad == cpu_grad.

rohan-varma replied:
I feel like the invariant is valuable because one of the goals of optimizer in backward is to cut down on memory usage by not holding on to sharded gradients that have already been applied on the parameter. 

Due to this, in GPU case, it seems essential that we set `handle.sharded_grad` (whether sharded_grad is the flat_param.grad or flat_param._saved_grad_shard) to None - if we don't do this and just set `orig_param.grad = None`, we won't see a memory reduction since FSDP internals will still hold on to a grad. 

in CPU case, it is a little trickier - ideally we'd like to set `_cpu_grad = None` as well, but this results in the extra reallocation in the next forward. Since we're in CPU memory, saving memory is less crucial because usually GPU memory << CPU memory, so maybe we just zero the cpu_grad instead of setting to None.

awgu replied:
Thanks for this analysis! I am fully onboard for the GPU case, but I am still skeptical about the performance on the CPU case (and I do not think right now that the CPU memory savings from freeing sharded gradients is meaningful). I am worried that the cost for this invariant is too high. Maybe we can go with this invariant first, but we can look at the performance (e.g. via traces) and reevaluate?

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -220,6 +220,15 @@ def _share_state_and_init_handle_attrs(
         attr_name_to_values[attr_name] = set()
     root_state._all_fsdp_states = traversal_utils._get_fsdp_states(root_module)
     root_state._all_handles = root_state._exec_order_data.all_handles  # share reference
+    # Update _has_optim_in_backward for each handle.
```
Let us add up-front assertion that `use_orig_params=True` to use `_in_backward_optimizers`. Then, we can simply `assert root_state._use_orig_params` and then remove the `if flat_param._params is not None`.


(`if flat_param._params is not None` is currently being used as a proxy for `use_orig_params=True` or `False`.)

awgu replied:
I see! That makes sense.

I think it could be clearer still to check `if handle._use_orig_params` instead of directly checking `flat_param._params is not None` and leveraging that `handle._use_orig_params == True` implies `flat_param._params is not None` since whether `use_orig_params=True` or not is the source of truth.

awgu commented on test/distributed/fsdp/test_fsdp_misc.py:
```diff
@@ -183,6 +184,96 @@ def _check_equal(local, fsdp):
+                        fsdp_overlap.named_parameters(), fsdp.named_parameters()
+                    ):
+                        pass
+                        # self.assertEqual(
```
Are these asserts meant to be skipped, or just for development?

rohan-varma replied:
these asserts need to pass.

rohan-varma commented:
another point for discussion: whether to disable `prepare_gradient_for_optim` call when overlapping optimizer. cc @awgu

rohan-varma committed:
Update on "FSDP optimizer overlap"


constraints:

1. No support for gradient accumulation
2. CPU offload is not supported. Will fix this for large scale CPU offload use cases as follow up.
3. Step is waited on in post backward final cb, when in theory it can wait
until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP optimizer overlap"


constraints:

1. No support for gradient accumulation
2. CPU offload is not supported. Will fix this for large scale CPU offload use cases as follow up.
3. Step is waited on in post backward final cb, when in theory it can wait
until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -883,6 +907,36 @@ def _post_backward_hook(
+                            # invariant that if flat_param.grad is None, then
+                            # each orig param grad is None.
+                            orig_param.grad = None
+                    # flat_param.grad is set to None for sharded strategies, set it here
+                    # for no_shard.
```
If this is the case, it could be worth including the `if not handle.uses_sharded_strategy:` to guard the `handle.flat_param.grad = None` to make this explicit.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -883,6 +907,36 @@ def _post_backward_hook(
+                    # similar to sharded_grad property.
+                    handle.flat_param.grad = None
+                    handle.flat_param._saved_grad_shard = None
+                    if handle._offload_params:
+                        handle.flat_param._cpu_grad = None
```
Relating to above comments, is this set to `None` only for the purpose of an invariant? What value do we get from doing this? In FSDP previously without optimizer overlap, I think we never set `_cpu_grad = None`?

(Feel free to only respond in one place.)

awgu reviewed (commented):
I did not read the test file yet since I have some questions about the main code.

rohan-varma commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1499,7 +1509,7 @@ def prepare_gradient_for_optim(self):
             # TODO (rohan-varma): test for full precision with keep_low_precision_grads
             if not self._force_full_precision and self._keep_low_precision_grads:
-                assert flat_param.grad is not None  # mypy
+                _p_assert(flat_param.grad is not None, "Unexpected None grad!")
```
It should not, but might be useful to keep to avoid introducing bugs in the future. I hit this assert when testing, and it was hard to isolate because we were not using `p_assert`

rohan-varma commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -857,7 +876,12 @@ def _post_backward_hook(
-                non_blocking = handle.uses_sharded_strategy
+                # TODO (rohan-varma): When CPU offload and optimizer overlap,
+                # non_blocking=True won't work since the copy may have not finished
+                # before the optimizer step executes on CPU.
```
> I vaguely recall that PyTorch inserts the synchronization to block anyway even if non_blocking=True

I did not know that, doesn't that defeat the point of non_blocking? 

From reading https://discuss.pytorch.org/t/should-we-set-non-blocking-to-true/38234/17 and https://forums.developer.nvidia.com/t/how-do-cpu-threads-know-that-gpu-kernel-is-finished/64598/2, it seems that if we copy from GPU to CPU with non_blocking=True, there's no easy way for the CPU to know copy is finished - we need to either synchronize or run with non_blocking=False. 

My current mental model is if we issue a GPU -> CPU copy with non_blocking=True, we need to synchronize before using the result on the CPU.

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_misc.py:
```diff
@@ -138,6 +139,7 @@ def test_fsdp_module_no_compute_grad(self, use_second_layer, sharding_strategy):
         # When use_second_layer=True, b is involved in forward computation but does
         # not receive grad in backward. Otherwise, b is not involved in forward
         # computation.
+
```
linter

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -1001,6 +1050,9 @@ def _post_backward_final_callback(
     root_state = state
 
     if root_state._sync_gradients:
+        # TODO (rohan-varma): this also waits for the overlapped optimizer step to finish
```
Should we track somewhere the follow-ups (e.g. an issue)?

rohan-varma replied:
@awgu tracking these here: https://github.com/pytorch/pytorch/issues/104479

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -857,7 +876,12 @@ def _post_backward_hook(
                 # TODO: Investigate why `NO_SHARD` breaks correctness when
                 # using `non_blocking=True` here.
-                non_blocking = handle.uses_sharded_strategy
+                # TODO (rohan-varma): When CPU offload and optimizer overlap,
```
Curious what the 'to-do' here is -- is a plan to use `non_blocking=True` still?

rohan-varma replied:
mainly to check if we do nonblocking=True and then sync before use on CPU, is it faster (my guess is probably not).

awgu commented on test/distributed/fsdp/test_fsdp_misc.py:
```diff
@@ -261,6 +263,151 @@ def _check_equal(local, fsdp):
+    @skip_if_lt_x_gpu(2)
+    def test_fsdp_optim_overlap_no_use_orig_params_error(self):
+
+        class MyModel(nn.Module):
```
nit: This module definition repeated across the two added unit tests could be de-duplicated. This unit test does not really depend on the model specifically, so we could similarly use a common model already defined.

awgu reviewed (approved):
Thanks for adding this!

There is a minor lint error. We should also follow-up on the re-allocating of a zeroed `_cpu_grad` and pinning it in case doing so has noticeable performance issues.

fegin commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -498,6 +498,16 @@ def __init__(
         self._debug_level = dist.get_debug_level()
         self._fully_sharded_module = fully_sharded_module
+        # NB: this is updated during lazy initialization because optimizers can be applied after
+        # wrapping with FSDP.
```
This seems to be a duplicated one as what is in `_runtime_utils.py`. Since optimizer may be applied after wrapping with FSDP, should we consolidate to only one check?

rohan-varma replied:
Consoildated to only after wrapping with FSDP check.

fegin commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -883,6 +907,31 @@ def _post_backward_hook(
+                        ):
+                            # TODO (rohan-varma): For CPU offload, this unfortunately
+                            # operates on CPU, because the parameters and gradients
+                            # have already been offloaded.
```
Is the plan to move the `step()` to GPU? If so, should we clarify the TODO here?

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

fegin reviewed (approved):
LGTM

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP optimizer overlap"



constraints:

1. No support for gradient accumulation
2. CPU offload runs step() on CPU. In future PRs ideally we'd run this on GPU.
3. When CPU offload + optimizer overlap, we have to copy the flat_param grad to CPU with non_blocking=False, otherwise step() might run on invalid data.
4. Step is waited on in post backward final cb, when in theory it can wait until the next forward.

Differential Revision: [D44809582](https://our.internmc.facebook.com/intern/diff/D44809582/)

[ghstack-poisoned]
